Simple inlining for old optimizer.

ekpyron opened the pull request:
References https://github.com/ethereum/solidity/issues/10698

ekpyron committed:
Add inlining for old optimizer.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,101 @@
+			}
+			lastTag.reset();
+		}
+		if (item.type() == Tag)
```
Shouldn't that be PushTag?

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,101 @@
+
+void Inliner::optimise()
+{
+	std::map<u256, AssemblyItems> inlinableBlocks;
```
Maybe use a vector view or something? The original m_items lives until the end.

ekpyron replied:
Yep - originally I wanted to modify the original one in place, but then noticed that's rather insane, but now views should indeed do!

ekpyron replied:
Ah, ok... then I have to copy items into ``newItems`` in the end instead of moving them, though... but copying ``AssemblyItem`` shouldn't be much more expensive than moving, so still better.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,101 @@
+
+	AssemblyItems newItems;
+	bool skipOne = false;
+	for (auto&& [item, nextItem]: ranges::views::zip(m_items, ranges::views::drop(m_items, 1)))
```
Interesting method! Could be simpler using views::sliding though.

ekpyron replied:
``sliding`` unfortunately gives me some vector or array or range or whatever it is and not a tuple, though... that's why I did it like this... but I'll have another look :-). Generally the ``skipOne`` thing is a bit ugly...

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,101 @@
+		}
+		if (
+			item.type() == PushTag &&
+			nextItem.type() == Operation &&
```
I think you might be able to do `nextItem == Instruction::JUMP` because of implicit constructors.

ekpyron replied:
``==`` seems to also compare ``data`` though...

ekpyron replied:
Ah no, it doesn't, nice!

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,101 @@
+	bool skipOne = false;
+	for (auto&& [item, nextItem]: ranges::views::zip(m_items, ranges::views::drop(m_items, 1)))
+	{
+		if (skipOne)
```
This mechanism might actually be a reason to use iterators instead...

ekpyron replied:
Yeah... I'll see if I can find something nicer or otherwise switch to iterators...

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `ab7f7681d393d633149a71a8c842f1f8faddaa1b`:
```
Coding style error:
 libevmasm/Inliner.cpp:76: for(auto it = m_items.begin(); it != m_items.end(); ++it)

```
Please check that your changes are working as intended.

stackenbotten commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,93 @@
+		return;
+
+	AssemblyItems newItems;
+	for(auto it = m_items.begin(); it != m_items.end(); ++it)
```
Coding style error

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,92 @@
+
+void Inliner::optimise()
+{
+	std::map<u256, ranges::span<AssemblyItem>> inlinableBlocks;
+
+	std::optional<size_t> lastTag;
+	for (auto&& [index, item]: m_items | ranges::views::enumerate)
+	{
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, true))
+		{
+			if (item == Instruction::JUMP)
+			{
+				ranges::span<AssemblyItem> items(m_items | ranges::views::slice(*lastTag + 1, index + 1));
+				if (isInlineCandidate(items))
+					inlinableBlocks.emplace(make_pair(m_items[*lastTag].data(), items));
+			}
+			lastTag.reset();
+		}
+		if (item.type() == Tag)
+			lastTag = index;
+	}
```
```suggestion
	std::map<u256, ranges::span<AssemblyItem>> inlinableBlocks = determineInlinableBlocks(_items);
}
```

ekpyron commented on libevmasm/Inliner.h:
```diff
@@ -0,0 +1,49 @@
+class Inliner
+{
+public:
+	explicit Inliner(AssemblyItems& _items, int _inlineMaxOpcodes = 6): m_items(_items), m_inlineMaxOpcodes(_inlineMaxOpcodes) {}
```
For hitting a function with two return values I had to increase to 6, the other tests were hit with 5 already...
I wonder what's a reasonable value here :-).

chriseth replied:
We might also take the "runs" parameter into account here.

chriseth replied:
Or just make it a setting...

ekpyron replied:
Made it a setting now - but the default value should maybe still depend on ``runs``...

ekpyron replied:
Or we go all-in for some proper heuristics, i.e. https://github.com/ethereum/solidity/pull/10761#discussion_r558287887

ekpyron commented on test/libsolidity/Assembly.cpp:
```diff
@@ -209,7 +209,10 @@ BOOST_AUTO_TEST_CASE(jump_type)
 			jumpTypes += item.getJumpTypeAsString() + "\n";
 
-	BOOST_CHECK_EQUAL(jumpTypes, "[in]\n[out]\n[in]\n[out]\n");
+	if (solidity::test::CommonOptions::get().optimize)
+		BOOST_CHECK_EQUAL(jumpTypes, "[in]\n[out]\n[out]\n[out]\n");
+	else
+		BOOST_CHECK_EQUAL(jumpTypes, "[in]\n[out]\n[in]\n[out]\n");
```
I hadn't really thought about this so far... but this might be a problem for debuggers...

chriseth commented on test/libevmasm/Optimiser.cpp:
```diff
@@ -1233,7 +1233,17 @@ BOOST_AUTO_TEST_CASE(jumpdest_removal_subassemblies)
-	main.optimise(true, solidity::test::CommonOptions::get().evmVersion(), false, 200);
+	Assembly::OptimiserSettings settings;
+	settings.isCreation = false;
+	settings.runInliner = false;
```
Why don't you run it?

ekpyron replied:
Because this is a test for the jumpdest remover...
I could run it, but then I'd need to adjust the expectations of this test as well :-). I could do that...

chriseth replied:
Ah sure, didn't see that!

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,322 @@
+		 * tag_1:
+		 * ...{B}...
+		 * PUSHTAG(someFunction)
+		 * JUMP(tag_2) // outOfBlock: into
```
```suggestion
		 * JUMP // outOfBlock: into
```

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,322 @@
+		 * ...{A}...
+		 * ...{B}...
+		 * PUSH(tag_2_out)
+		 * JUMP(tag_2) // result: ordinary
```
Why is this not "out of"?

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,322 @@
+				if (auto const *inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
+				{
+					newItems += *inlinableBlock;
+					newItems.back().setJumpType(determineJumpType(nextItem.getJumpType(), inlinableBlock->back().getJumpType()));
```
I guess the main problem is when the breaking instruction is not a jump instruction - you cannot really set a jump type on a call opcode.

ekpyron replied:
It will always be a jump (as I just explained on gitter and as seen in determineInlinableBlocks) :-).

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `79af73579c048cbf45f3ab7b4ca1954f66f45334`:
```
Coding style error:
 libevmasm/Inliner.cpp:199: CMakeLists.txt CODE_OF_CONDUCT.md CODING_STYLE.md CONTRIBUTING.md Changelog.md LICENSE.txt README.md ReleaseChecklist.md SECURITY.md cmake codecov.yml docs libevmasm liblangutil libsmtutil libsolc libsolidity libsolutil libyul scripts snap solc test tools ...{B}...

```
Please check that your changes are working as intended.

stackenbotten commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,254 @@
+		 *    leave // intoBlock: out of
+		 * }
+		 * function g() {
+		 * 	  ...{B}...
```
Coding style error

ekpyron replied:
C'mon, it's a comment...

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `5d4ce5eb1baf5ebab5186f5f0d010e689038537d`:
```
Coding style error:
 libevmasm/Inliner.cpp:193: CMakeLists.txt CODE_OF_CONDUCT.md CODING_STYLE.md CONTRIBUTING.md Changelog.md LICENSE.txt README.md ReleaseChecklist.md SECURITY.md cmake codecov.yml docs libevmasm liblangutil libsmtutil libsolc libsolidity libsolutil libyul scripts snap solc test tools ...{B}...

```
Please check that your changes are working as intended.

stackenbotten commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,248 @@
+		 *    leave // intoBlock: out of
+		 * }
+		 * function g() {
+		 * 	  ...{B}...
```
Coding style error

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `621235dc651ebcffd40633a7d172e0610d1bc0ae`:
```
Coding style error:
 libevmasm/Inliner.cpp:83: switch(jumpTypeToInt(_intoBlock) + jumpTypeToInt(_outOfBlock))

```
Please check that your changes are working as intended.

stackenbotten commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,125 @@
+		}
+		assertThrow(false, OptimizerException, "");
+	};
+	switch(jumpTypeToInt(_intoBlock) + jumpTypeToInt(_outOfBlock))
```
Coding style error

ekpyron commented:
So this will affect jump labels and we need input in how best to deal with that. So first a summary of what is happening here:

With this PR, whenever there is a JUMP(tag) to a tag at which there are only a few opcodes in straight, simple control-flow followed by another JUMP, it will remove the jump to the tag and copy the opcodes up to that JUMP in its place.

If both the first jump we replace and the jump at the end of the block we replace it with are "Ordinary" jumps, then the resulting jump is also just an "Ordinary" jump and nothing special happens, but if one or both of them were jumps into- or out-of a function, things get more complicated.

The following situations can occur:

| Original Jump to be replaced | Jump at end of Block originally jumped to | Jump in resulting inlined code | Comment
|----------------------------------------|--------------------------------------------------------|-----------------------------------------|---------------
| Ordinary                                 | Ordinary                                                     | Ordinary                                  | Nothing special but one jump less.
| Ordinary                                 | Into Function                                              | Into Function                            | Nothing special but one jump less.
| Ordinary                                 | Out of Function                                          | Out of Function                        | Nothing special but one jump less.
| Into Function                          | Ordinary                                                     | Into Function                            | This means there will be source locations from inside the function before the jump into the function
| Into Function                          | Into Function                                              | WILL NOT INLINE                 | We could inline here, but this would change the balance of Into-Function and Out-of-Function jumps - i.e. the situation is that the first thing ``f`` does is to call ``g``, originally we have a jump to ``f``, then a jump to ``g``, but now we directly jump to ``g``, but we will later still see *two* "Out of function" jumps, misaligning the balance.
| Into Function                          | Out of Function                                          | Ordinary                              | This is "traditional" inlining of an entire function. This means in a code block source locations of a function will occur without any jump into or out of this function
| Out of Function                      | Ordinary                                                     | Out of Function                            | This means there will be source locations from outside the function at the end of the function before the jump out.
| Out of Function                      | Into Function                                               | Ordinary                                      | This can happen if two functions are called directly after each other. Instead of jumping out of one and into the other, we jump directly from one to the other. This means source locations will flip from one function to another function at an ordinary jump. In total the Jump-Into-Jump-Out-Of balance will be preserved, though.
| Out of Function                      | Out of Function                                               | WILL NOT INLINE                 | Analog to the Into-Function-Into-Function case. Can happen if the last thing a function does is calling another function. Instead of first returning from one and then from the other, one jump returns from both. We could inline here, but also not without disturbing the "Jump-Into-Jump-out-of"-total-balance.


So the first question is: how much of a problem is this for tooling already :-)?
Further questions would be: How much bigger of a problem would this be if we also inlined in the "Into-Into"-Function and "Out-of-Out-of"-Function cases?
Is there anything better we can do?

Generally, after this PR the function begining or end will no longer necessarily be tied to jumps. Actually, though, I'm not sure why those jump labels are actually needed - wouldn't looking at the source locations before and after a jump have been enough before a PR? If so that would mean looking at source locations would still be sufficient after this PR, but there's probably more to it - maybe the stack layout of arguments and return variables? This will only be guaranteed to match the expected layout between the "jump-in" and "jump-out" jumps after this PR, independently of whether source locations are already inside the function - and in the "out-of-into"-case the stack layout would change at an ordinary jump inside the function, at which the layout switches from the usual one from one function to the one of the next. Would it be more helpful to still mark this as jump "into" the next function, even though there will only be one jump-out for two jump-in's this way?

So the optimizer step itself should work and be correct, but we're not sure how to deal with this part :-).

ekpyron commented:
Another note: only inlining in the "ordinary-*" cases (they should not affect tooling at all) and in the "into-out-of" case, i.e. inlining an entire function in place, would already make this step quite useful and solve the loop increment issue that triggered this.
So the first important question is whether *that* would cause tooling issues; the rest can have lower priority and easily postponed until after 0.8.1 and discussed further.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `eb4252c3940741ac30a1d6594ab1168792c9355f`:
```
Coding style error:
 libevmasm/Inliner.cpp:75: for(auto&& [tag, items]: inlinableBlockItems)

```
Please check that your changes are working as intended.

stackenbotten commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	map<u256, InlinableBlock> result;
+	for(auto&& [tag, items]: inlinableBlockItems)
```
Coding style error

ekpyron commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,143 @@
+bool Inliner::isInlineCandidate(InlinableBlock const& _block) const
+{
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+	// Always try to inline if there is at most one call to the block.
+	if (_block.pushTagCount == 1)
+		return true;
```
The problem the way this is written for now is that if we're unlucky, we'll already have the one block pushing this tag ten times before this is called... maybe that's worth accounting for - should be possible by tracking all ``PushTag``s we've already inlined...

ekpyron replied:
Should be fine now.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `a0a06a7d9c44c5cdc57d47a305f742237a72c5ec`:
```
Coding style error:
 libevmasm/Inliner.cpp:45: for(AssemblyItem const& item: _block.items)

```
Please check that your changes are working as intended.

stackenbotten commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,150 @@
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+
+	// Never inline tags that reference themselves.
+	for(AssemblyItem const& item: _block.items)
```
Coding style error

ekpyron commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+}
+}
+
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
```
In this function we should have all information available now, that we may need for any kind of complex heuristics I think.

We have the jump to the tag that we want to inline (and thereby its jump type), we know all the opcodes to be inlined (including the exit jump and its jump type), and we know how often the tag is pushed throughout the bytecode.

So we could actually now really calculate the cost of inlining vs not inlining with a proper gas estimate using the ``runs`` parameter...

So we have to decide if we want to do this right away or keep it simple for now - keeping in mind that this might affect which standard-json options we want to define for it.

ekpyron replied:
One thing we could actually additionally determine by walking the items one more time up front: how many of the ``PushTag``s we've seen actually have the form ``PUSHTAG JUMP``, i.e. definitely can *actually* be inlined...
Although we could still miss ``PUSHTAG PUSH SWAP1 JUMP`` for example... anyways, it's probably enough as it stands for a decent heuristics.

chriseth replied:
Inlining can result in further optimizations, so computing the actual gas tradeoff is not perfect, but could certainly help.

aarlt commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+		{
+			InlinableBlock block{items, *numPushes};
+			if (isInlineCandidate(tag, block))
+				result.emplace(std::make_pair(tag, InlinableBlock{items, *numPushes}));
```
Isn't it possible to just use `block` here instead of `InlinableBlock{items, *numPushes}`

haltman-at commented:
> So the first question is: how much of a problem is this for tooling already :-)?

Hm, so, Truffle Debugger already has lots of problems with optimized code; I can't speak for writers of other tools, of course.  So yeah this might cause problems for locating variables but, like, every optimization does.  I could go into detail about exactly how it might make things worse, but I'll skip that for now I think.

I imagine the bigger problem is the question of to what extent this would mess up stack traces, since there are a larger number of tools out there that do that (and see also #10715 for a different example of a recent way the optimizer can mess up stack traces), and that's something people might expect to work a bit more robustly.

So, let's go through the cases here and see what problems this would cause for stacktracing (let's imagine that the code in that segment reverts for some reason).

* Case `--`: No problem here, obviously.
* Case `-i`: Ditto.
* Case `-o`: Ditto.
* Case `i-`: The source location will show a point in the function called, but there won't be a stackframe for it, it'll still show the caller.  In particular, it won't show where in the caller the call happened.  This is... not great.  Exactly how bad is it?  I don't know, maybe someone else can answer that better.
* Case `ii`: Not inlined, so no problem. :)  If it were to be inlined, it would have the problem of case `i-`, with the additional problems caused by unbalancing (that could mess up strack traces for reverts occuring anywhere later in the transaction).
* Case `io`: Similar to case `i-`.
* Case `o-`: The source location will show a point in the function returned to, but the called function won't have been popped from the stack yet.  This doesn't strike me as as big a problem as the problem of `i-`, since while it might be a little confusing, there's not really the same level of information loss.
* Case `oi`: This case has the problems of case `o-`, but also has the additional problem that any reverts from within the second function call will show as coming from the first function call (although of course they'll still have correct source location), with nothing to show where the common caller made that second function call.  Blech.  So this is case kind of combines the problems of `o-` with the problems of `i-`, except that the latter is worse than in case `i-`, since there it only applies within a small segment, whereas here it applies for the entire duration of the second function call.
* Case `oo`: Not inlined, so no problem. :)  If it were to be inlined, it would have the problems of case `o-`, with the additional problems caused by unbalancing (that could mess up strack traces for reverts occuring anywhere later in the transaction).

I guess the question of how much of a problem all this is is to a substantial extent a question of how well people expect tools to work with otimized code.

> Further questions would be: How much bigger of a problem would this be if we also inlined in the "Into-Into"-Function and "Out-of-Out-of"-Function cases?

Is there a way to make sure you only do those in matching pairs around the same call?

Otherwise, unbalancing things is going to result in bad stack traces if the revert happens anywhere after the unbalance, in ways that are probably worse than the problems I described above.

I mean, unless you want to add a way to indicate "this jump corresponds to n jumps in/out" instead of just one. :P

> Is there anything better we can do?

> Would it be a helpful replacement of jump labels to mark arbitrary instructions instead of jumps as "into"- and "out-of"-functions?

(going a bit out of order here :) )

I'd be all for that!  Indeed I think I've asked about this before, for the purpose of marking the beginnings and endings of functions that don't currently use jumps, such as constructors and fallback functions.  (And there's the whole mess with modifiers, but *those* probably shouldn't use such markings, because those are more complicated.  I don't really have a great idea how to handle those; some sort of separate "start of modifier" marking...?)

Of course, doing this might break tooling that isn't ready for it.  But I personally am all for it.  Adding support for it in Truffle Debugger wouldn't be hard.  (I imagine adding support for it in other tools wouldn't be hard either...)

> Generally, after this PR the function begining or end will no longer necessarily be tied to jumps. Actually, though, I'm not sure why those jump labels are actually needed - wouldn't looking at the source locations before and after a jump have been enough before a PR?

I will say that jump labels are definitely very helpful.  Trying to go based on comparing source locations... well, it might be doable, but it'd have a number of caveats.  (Just off the top of my head, one would have to make sure that all of this worked correctly with modifiers, with base constructors, with getters, with initializers, with unmapped code...)  I mean even just determining which function a given source range falls in isn't trivial, you know?  Truffle Debugger might be able to handle this, given all the crazy stuff it already does, but I expect writers of other tools, who are probably trying to keep them simple, rather than monstrosities of special cases, would really not appreciate such a change.

I would also in general expect such a method to not necessarily work for optimized code, whose source mappings seem to often work in unexpected ways.

> If so that would mean looking at source locations would still be sufficient after this PR, but there's probably more to it - maybe the stack layout of arguments and return variables? This will only be guaranteed to match the expected layout between the "jump-in" and "jump-out" jumps after this PR, independently of whether source locations are already inside the function - and in the "out-of-into"-case the stack layout would change at an ordinary jump inside the function, at which the layout switches from the usual one from one function to the one of the next. Would it be more helpful to still mark this as jump "into" the next function, even though there will only be one jump-out for two jump-in's this way?

Truffle Debugger at least doesn't use jump markings for this purpose, so that at least would not be a problem.  Mind you, it gets variable locations wrong all the time when dealing with optimized code, so perhaps I should instead say it wouldn't be any more of a problem, but the point is, it wouldn't be any more of a problem.

Although, while it doesn't use jump markings for determining where variables are located, it does use them for determining what variables are active.  I don't think this should cause *too* much of a problem, but there would at least be some problems there.  For instance, if a function `f` made multiple inlined calls to `g`, then on the second inlined call to `g`, its variables would show up, with incorrect values, before they were declared.  Similarly, if instead of the calls to `g` being inlined, there was a case `oi` where the return to `f` and then second jump into `g` were what was inlined, the same problem would occur.

Oh wait, actually, `i-` would cause a problem too.  Since if this happens when `f` calls `g`, this could possibly cause early variables of `g` declared in the inlined segment (like, say, the parameters) to get associated with the wrong stack frame, and so they wouldn't display for the duration of `g`.  Blech.  I can at least think of a way to possibly fix that one, unlike the cases above, but, oy.  And actually I suppose the same thing could happen with `o-`, if any variables got declared in the inlined segment there.  That case doesn't seem fixable, unfortunately.

So, those are the problems that Truffle Debugger would have with it.

Obviously, I can't speak for what Remix does or whether it might mess their debugger up, or any other debuggers.

Unbalancing jump-ins and jump-outs seems really bad, though, as mentioned above, for anyone that's just trying to make stack tracing work, let alone anything more complicated.

chriseth commented:
So as a summary, @haltman-at , does the following work?
 - it is optimized code, so it is fine if some metadata gets lost, as long as it is not too bad.
 - as long as we do not destroy the in/out balance, it's probably fine, even though some variable information might get lost
 - let us think of better debugging annotations we can use in the future.

What do the others think?

haltman-at commented:
> So as a summary, @haltman-at , does the following work?
> 
>     * it is optimized code, so it is fine if some metadata gets lost, as long as it is not too bad.
> 
>     * as long as we do not destroy the in/out balance, it's probably fine, even though some variable information might get lost
> 
>     * let us think of better debugging annotations we can use in the future.
> 
> 
> What do the others think?

That seems like a fair summary, yes.  I'd be more worried about the effects on stack traces than the effects on variables.

chriseth commented on docs/metadata.rst:
```diff
@@ -80,6 +80,8 @@ explanatory purposes.
           details: {
             // peephole defaults to "true"
             peephole: true,
+            // maximum inline size defaults to 6
```
Disparity between commend and code.

chriseth commented on docs/using-the-compiler.rst:
```diff
@@ -243,6 +243,9 @@ Input Description
             // The peephole optimizer is always on if no details are given,
             // use details to switch it off.
             "peephole": true,
+            // Maximum inline size is 6, if no details are given,
```
Same here.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+
+	// Never inline tags that reference themselves.
+	for (AssemblyItem const& item: _block.items)
+		if (item.type() == PushTag)
+			if (_tag == item.data())
+				return false;
```
```suggestion
		if (item.type() == PushTag &&_tag == item.data())
			return false;
```

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+	{
+		// The number of PushTag's approximates the number of calls to a block.
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
```
Note that at construction time, the data might contain both the construction and post-deploy tag. In that case, though, it is probably always a dynamic jump and can be ignored.

ekpyron replied:
I'm still wondering whether we should handle that more properly, though...

chriseth replied:
In general, do you handle dynamic references? Can you add some tests?
```
contract C {
  function() internal pure returns (uint) x;
  function f() internal pure returns (uint) { return 6; }

  function g() public { x = f; }
  function h() public returns (uint) { return x(); }
  function i() public returns (uint) { return f(); } // this should be inlined
}
```
And another test where the assignment to `x` is done in the constructor and another where `h` and `i` are called in the constructor.

ekpyron replied:
So far I don't - but maybe we should... but I actually only conceptually know how it works, I'll need to look at it in more detail... we'd probably use ``AssemblyItem::splitForeignPushTag`` and just consider the jumps into the same ``subId``? It'd be nice to know at deployed code optimization, if a tag was referred to by constructor code and change the heuristics to assume the function isn't removable in that case, but not sure how easy that is...

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, true))
```
Actually, the `_msizeImportant` flag can be safely set to `false`, can't it? If we set it to `false`, we can also inline calls to `keccak` and `mload`.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
```
Maybe also comment: We use `breaksCSEAnalysisBlock` here so that CSE analysis can hopefully optimize away the return jump.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, true))
+		{
+			if (item == Instruction::JUMP)
+				inlinableBlockItems.emplace(
+					piecewise_construct,
+					forward_as_tuple(_items[*lastTag].data()),
+					forward_as_tuple(_items | ranges::views::slice(*lastTag + 1, index + 1))
+				);
```
Why is this so complicated?
```suggestion
				inlinableBlockItems[_items[*lastTag].data()]= _items | ranges::views::slice(*lastTag + 1, index + 1);
```

ekpyron replied:
Because if I do it often enough, maybe people will get used to it as the canonical way for zero-cost in-place construction of map elements, which it is ;-P.

chriseth replied:
They won't ;)

ekpyron replied:
Well, ok, I'm changing it... but I still must say: you are aware that this construct *always* works and is *always* optimal, and I'd say once you have seen it only once, it's also absolutely clear what it does...
Whereas the (IMHO only seemingly) simpler ``...[...] = ...;`` has to construct and - if you're unlucky - copy two temporaries, even before that default-construct an empty element to be copied (or in the best case moved) over, which may not even be possible, if you don't have a default-constructor, it gives no feedback about it already having been there or not, etc.

So I think one can very well argue that ``...[...] = ...`` is *always* inferior - if both work, in-place constructing is always cheaper, and there are cases in which *only* inplace-constructing will even work.

But yeah, on the other hand, it's not important enough to argue, so having said all that, I'll still just change it ;-).

chriseth replied:
It's not inferior, it is so much better with regards to readability. And reduced readability can easily lead to bugs. I would always prefer a language with syntax over a language without syntax.
```(for (def i) (gt i 0) (assign i (add i 1)) (mstore i (mul i 2)))```
Which elements of the for loop belong where?

ekpyron replied:
I don't think that's analog in any way - ``container.emplace(piecewise_construct, `` exactly says what it does - it piecewise constructs something into my container. And ``forward_as_tuple(keyArgs...)`` and ``forward_as_tuple(valueArgs...)`` exactly say how the stuff is constructed - using forwarded arguments.
On the other hand a whole bunch of hidden things happen in ``container[keyArgs] = valueArgs;``. First I construct a temporary key object - then I potentially default-construct a value object in the container, then I construct a temporary value type, then I move or copy that value type into the container. Loads of stuff no-one might ever suspect. And every step along the way may be undesirable or may just not work...

ekpyron replied:
But well, I did change it, so not much point arguing about it ;-).

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+		return;
+
+	AssemblyItems newItems;
+	for (auto it = m_items.begin(); it != m_items.end(); ++it)
```
Actualyl can't we use `iterateReplacingWindow` here?

ekpyron replied:
``iterateReplacingWindow`` modifies the ``vector`` we iterate over in-place - so that would invalidate the ``ranges::span``s into ``m_items`` here... so yes, we could use it, but then we would need to copy the items into ``InlinableBlock`` instead of only storing ``span``s into ``m_items`` there. Both fine with me...

chriseth replied:
Ah right, then it is probably not a good idea.

yann300 commented:
Sorry for the late answer.
Here are some early feedback:

> Generally, after this PR the function begining or end will no longer necessarily be tied to jumps

That will be a problem for us ( `i-`, `o-` ) because jump in function are at the moment also identified using source location and JUMP mark. So I am pretty sure that will break our analyzer. But might be good time for us to switch to a better solution.
`jump labels to mark arbitrary instructions instead of jumps as "into"- and "out-of"-functions` will just solve the issue.

As far as I can see, `io` could also be an issue. I m not sure how we will be able to know when we are stepping inside a function, if there's no `into-`, `out-of` tags.
`oi` also, that's vital to be aware of stepping out of a function.

> - wouldn't looking at the source locations before and after a jump have been enough before a PR?

Not really, well that could work, but jump label is for us far more accurate.

> Would it be a helpful replacement of jump labels to mark arbitrary instructions instead of jumps as "into"- and "out-of"-functions?

This could solve all the issues you've mentioned. Would be important I think to include `mark arbitrary instructions` in the release that will ship this PR and to give us time to integrate the changes.

bshastry commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+	std::optional<size_t> lastTag;
+	for (auto&& [index, item]: _items | ranges::views::enumerate)
+	{
+		// The number of PushTag's approximates the number of calls to a block.
```
```suggestion
		// The number of PushTags approximates the number of calls to a block.
```

hrkrshnn commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,182 @@
+		{
+			InlinableBlock block{items, *numPushes};
+			if (isInlineCandidate(tag, block))
+				result.emplace(std::make_pair(tag, InlinableBlock{items, *numPushes}));
```
Can't we remove `make_pair` here?

hrkrshnn commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,178 @@
+		{
+			AssemblyItem const& nextItem = *next(it);
+			if (item.type() == PushTag && nextItem == Instruction::JUMP)
+				if (auto *inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
```
```suggestion
				if (auto inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
```
Isn't it redundant? Fine with both situations. If you keep it, then `auto*` followed by whitespace, right? If you remove it, then similarly for line 148.

ekpyron replied:
It's redundant, but in the constant fight about overusing ``auto`` or not it often worked as a compromise ;-).

ekpyron commented:
After some more experimenting with this I now restricted this PR to the "io" case (i.e. inlining entire functions) only. That's the most beneficial case and the one I'd expect to be least destructive for debuggers (worst case is the debugger will be confused for a few instructions, but should hopefully be able to recover). The second-most-beneficial case I found to be "ii", but since that will break the balance of stack frames, we can postpone it until we have better more generalized debugging output for this.

CI is still building soljson.js for this current version. A canonical test case for checking debugger behaviour would be https://github.com/ethereum/solidity/blob/21b9a22652943f45a5de36db88ad39cc850ecb3f/test/cmdlineTests/optimizer_inliner_inc/input.sol - ``unsafe_inc`` will be inlined with the changes in this PR.

bshastry commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,160 @@
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
```
Just curious: What is a simple example of non-straight control flow in EVM terms?

ekpyron replied:
E.g. ``CALL``, ``JUMPI``, etc.

bshastry commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,160 @@
+			lastTag = index;
+	}
+
+	// Filter candidates for general inlinability and store the number of PushTag's alongside the assembly items.
```
```suggestion
	// Filter candidates for general inlinability and store the number of PushTags alongside the assembly items.
```

bshastry commented on test/libevmasm/Optimiser.cpp:
```diff
@@ -1467,6 +1478,91 @@ BOOST_AUTO_TEST_CASE(cse_replace_too_large_shift)
+		Instruction::CALLVALUE,
+		Instruction::SWAP1,
+		jumpOutOf,
+	};
```
Note to self: This is

```
Push [Tag] 1
Push [Tag] 2
JUMP
Tag_1
STOP
Tag_2
CALLVALUE
SWAP1
JUMP
```

bshastry replied:
Does this mean the following?

```
1. Jump to tag_2 setting return location to tag_1
2. Issue callvalue, SWAP1, and jump back to return location previously set
3. Stop
```

ekpyron replied:
Kind of - not sure I read correctly what you're points are :-).
The first "Push [Tag] 1" pushes the location of tag_1 as return address to the stack. ``Push [Tag] 2`` pushes the location of tag_2 as jump destination. So this will jump to ``Tag_2`` with ``Tag_1`` still on the stack. ``CALLVALUE`` is then just something the imagined function here does, i.e. push the callvalue to the stack (simulating a function return value). ``SWAP1`` rearranges the swap to move the return address to the stack top, then the ``JUMP`` jumps back to ``Tag_1``.

bshastry commented on test/libevmasm/Optimiser.cpp:
```diff
@@ -1467,6 +1478,91 @@ BOOST_AUTO_TEST_CASE(cse_replace_too_large_shift)
+		AssemblyItem(Tag, 2),
+		Instruction::CALLVALUE,
+		Instruction::SWAP1,
+		jumpOutOf,
```
NTS: This is

```
CALLVALUE
SWAP1
JUMP
Tag_1
STOP
Tag_2
CALLVALUE
SWAP1
JUMP
```

bshastry replied:
Does this mean the following?

```
1. Callvalue, SWAP1, and jump to location specified by stack top before callvalue was issued (return addr is value returned by callvalue)
```

Sorry a bit confused how to read this. Dunno if I am reading it correctly in the first place :-)

ekpyron replied:
This is the bare "inlined" version of the above.
``PushTag(1)`` still pushes the return point for the following jump. Then ``CALLVALUE`` is the "return value" of the inlined function. ``SWAP1`` rearranges the stack for jumping, s.t. the following jump jumps to ``Tag_1``.
So in both cases you end up at ``STOP`` with the callvalue on stack - i.e. the code is equivalent.

In practice there are more optimizer steps involved. In particular CSE will move the ``PushTag(1)`` down towards the jump, eliminating the SWAP1, i.e. the next step would be:

```
		Instruction::CALLVALUE,
		AssemblyItem(PushTag, 1),
		Instruction::JUMP,
		AssemblyItem(Tag, 1),
		Instruction::STOP,
		AssemblyItem(Tag, 2),
		Instruction::CALLVALUE,
		Instruction::SWAP1,
		jumpOutOf,
```
Then the peephole optimizer removes ``Push(Tag_1) jump tag_1``, i.e. you get
```
		Instruction::CALLVALUE,
		AssemblyItem(Tag, 1),
		Instruction::STOP,
		AssemblyItem(Tag, 2),
		Instruction::CALLVALUE,
		Instruction::SWAP1,
		jumpOutOf,
```
Furthermore ``tag_1`` and  ``tag_2``  would now be unused and be removed and the code behind ``tag_2`` removed, because it's unreachable, yielding:
```
		Instruction::CALLVALUE,
		Instruction::STOP
```
in the end :-).

Most of this has nothing to do with this PR, though.

bshastry replied:
Ah I forgot to read the first push tag instruction in the optimised sequence :)

hrkrshnn commented on libevmasm/Inliner.h:
```diff
@@ -0,0 +1,58 @@
+	bool isInlineCandidate(u256 const& _tag, InlinableBlock const& _block) const;
+	/// @returns the exit jump for the block to be inlined, if a particular jump to it should be inlined, otherwise nullopt.
+	std::optional<AssemblyItem> shouldInline(u256 const& _tag, AssemblyItem const& _jump, InlinableBlock const& _block) const;
+	std::map<u256, InlinableBlock> determineInlinableBlocks(AssemblyItems const& _items) const;
```
Could use documentation.

hrkrshnn commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,149 @@
+		// Using breaksCSEAnalysisBlock will hopefully allow the return jump to be optimized after inlining.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, false))
+		{
+			if (item == Instruction::JUMP)
```
Can't we check here that the item's jump type is `OutOfFunction`?

ekpyron replied:
That's not the only kinds of jumps we can theoretically inline (even though we will only inline those for now).

hrkrshnn replied:
Are you planning to keep it like this, or will you add an extra check on the jump type?

hrkrshnn commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,149 @@
+	{
+		exitJump.setJumpType(AssemblyItem::JumpType::Ordinary);
+		size_t codeSize = static_cast<size_t>(_block.items.size()) - 1;
+		bigint uninlinedCost = bigint(m_runs) * 24u * _block.pushTagCount + (8u * _block.pushTagCount + 2 + codeSize) * 200u;
```
Can you write a comment about the formula?

ekpyron replied:
Given the newfound review eagerness (which is a good thing ;-)), I probably should have drafted the PR for this. It wasn't meant to stay like this anyways.

hrkrshnn commented on libevmasm/Inliner.h:
```diff
@@ -0,0 +1,58 @@
+ */
+#pragma once
+
+#include <libsolutil/Common.h>
+#include <range/v3/view/span.hpp>
```
```suggestion
#include <libsolutil/Common.h>

#include <range/v3/view/span.hpp>
```

ekpyron replied:
We really don't have a fixed style for that, do we?

ekpyron replied:
I can add a linebreak, though, sure.

hrkrshnn commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,149 @@
+						for (AssemblyItem const& inlinedItem: inlinableBlock->items)
+							if (inlinedItem.type() == PushTag)
+								if (auto* block = util::valueOrNullptr(inlinableBlocks, inlinedItem.data()))
+									++block->pushTagCount;
+
```
I don't really follow how `pushTagCount` is used. Can you please explain?

ekpyron replied:
``pushTagCount`` counts how many times the tag at which the respective inline candidate lives is pushed, i.e. roughly how many jumps there are into that block. When inlining we can increase the number of pushes, though, that's accounted for here.
Unlikely to happen currently, since we restrict to full function inlining, but more likely to happen in the more general case and still not impossible in full function inlining.

ekpyron commented:
FYI this is the difference in gas expectations https://gist.github.com/ekpyron/0b3c2b0ac499dfad578ad01ecdfd34ee stemming from rebasing this on top of https://github.com/ethereum/solidity/pull/10863 (the diff is between updating once with the inliner disabled and once with it enabled).

chriseth commented on Changelog.md:
```diff
@@ -4,6 +4,7 @@ Language Features:
 
 
 Compiler Features:
+ * Optimizer: Simple inlining of functions with straight control flow.
```
Should we explain this in more depth in the documentation?

ekpyron replied:
Maybe, yeah...

chriseth commented on docs/using-the-compiler.rst:
```diff
@@ -246,6 +246,9 @@ Input Description
             // use details to switch it off.
             "peephole": true,
+            // The inliner is always on if no details are given,
+            // use details to switch it off.
```
Is this still up to date?

ekpyron replied:
How do you mean? It's what the PR does, isn't it?

chriseth replied:
Ah, this is in the details, ok!

chriseth commented on libevmasm/Assembly.cpp:
```diff
@@ -421,6 +423,12 @@ map<u256, u256> Assembly::optimiseInternal(
 
+		if (_settings.runInliner)
+		{
+			Inliner inliner{m_items, _settings.expectedExecutionsPerDeployment, _settings.isCreation, _settings.evmVersion};
```
Maybe just pass the settings as one object?

chriseth replied:
Do you consider `runs` to be `1` for `isCreation` being true?

ekpyron replied:
> Maybe just pass the settings as one object?

Not without pulling in another header in the optimizer run (which seemed to intentionally only forward-declare ``AssemblyItem`` and the settings class is a subclass of that - I could pull it out of the class and forward-declare it as well though).



> Do you consider `runs` to be `1` for `isCreation` being true?

Yep - the settings are automatically adjusted like that for creation, i.e. ``_settings.expectedExecutionsPerDeployment`` should be ``1``, if ``isCreation`` is true.

chriseth commented on libevmasm/Inliner.h:
```diff
@@ -0,0 +1,65 @@
+{
+public:
+	explicit Inliner(AssemblyItems& _items, size_t _runs, bool _isCreation, langutil::EVMVersion _evmVersion)
+	: m_items(_items), m_runs(_runs), m_isCreation(_isCreation), m_evmVersion(_evmVersion) {}
```
```suggestion
	explicit Inliner(AssemblyItems& _items, size_t _runs, bool _isCreation, langutil::EVMVersion _evmVersion):
		m_items(_items), m_runs(_runs), m_isCreation(_isCreation), m_evmVersion(_evmVersion) {}
```

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+{
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+
+	// Never inline tags that reference themselves.
```
This is a condition that does not lead to invalid code when violated, does it? (I'm wondering because comparing tags by their data might be tricky at construction time)

ekpyron replied:
No, bad it'll blow up code...
```
tag_1:
  CALLVALUE
  PUSHTAG(tag_1)
  JUMP
```

after one excution:

```
tag_1:
  CALLVALUE
  CALLVALUE
  PUSHTAG(tag_1)
  JUMP
```

and so on exponentially...

But: since we restrict to only jump-in jump-out this will never happen anyways. And even if we were to enable more combinations, it'd always be infinite-loop situations... so not that dramatic...

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+
+	// Never inline tags that reference themselves.
+	for (AssemblyItem const& item: _block.items)
+		if (item.type() == PushTag && _tag == item.data())
```
```suggestion
		if (item == AssemblyItem{PushTag, _tag})
```

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+		if (uint64_t const* numPushes = util::valueOrNullptr(numPushTags, tag))
+		{
+			InlinableBlock block{items, *numPushes};
+			if (isInlineCandidate(tag, block))
```
This can actually be checked during the first loop - would that be more readable?

ekpyron replied:
The idea was that ``isInlineCandidate`` might want to filter based on ``numPushes`` (i.e. throw away large blocks, unless only pushed once...) - but with the new heuristics that's probably obsolete, so I can move it, yep.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+	return result;
+}
+
+namespace {
```
```suggestion
namespace
{
```

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+template<typename RangeType>
+uint64_t codeSize(RangeType&& _itemRange)
+{
+	// TODO: what is the best choice for the ``_addressLength`` argument here?
```
2 is good, it is the max value.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+}
+}
+
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
```
Remove the first parameter?

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
+{
+	AssemblyItem exitJump = _block.items.back();
+
```
assert it is a jump?

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+	if (
+		_jump.getJumpType() == AssemblyItem::JumpType::IntoFunction &&
+		exitJump.getJumpType() == AssemblyItem::JumpType::OutOfFunction
+	)
```
return early.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+			AssemblyItem{Instruction::JUMP},
+			AssemblyItem{Tag}
+		};
+		static AssemblyItems const uninlinedJumpSitePattern = {
```
```suggestion
		static AssemblyItems const uninlinedFunctionPattern = {
```

chriseth replied:
(jump site could both mean the site to jump from or to jump to)

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+
+		// If the estimated runtime cost over the lifetime of the contract plus the deposit cost in the uninlined case
+		// exceed the inlined deposit costs, it is beneficial to inline.
+		if (bigint(m_runs) * uninlinedExecutionCost + uninlinedDepositCost > inlinedDepositCost)
```
Does this assume we can always delete the function if it is inlined?

ekpyron replied:
Yes... which is of course wrong in general... I could also calculate under the assumption that we *can't* remove it. It would probably also be possible to be sure about it (not only count pushtags, but also count which of those are ``pushtag jump``), but only for "right now" and not for "maybe after further optimization cycles"...

chriseth replied:
Can you take references from the surrounding assembly into account? That will probably not be removable since it is a dynamic jump.

ekpyron replied:
Yeah... let me see if I can... with "surrounding assembly" you mean the assembly the current assembly is a subassembly of, right :-)?

ekpyron replied:
Ah, ok, there is ``_tagsReferencedFromOutside`` available, I'll just have to pass that through, nice! I'll do that then!

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+						newItems += ranges::views::drop_last(inlinableBlock->items, 1);
+						newItems.emplace_back(*exitJump);
+
+						// We are removing one push tag to the block we inline.
```
Does this change the heuristic while are inlining?

ekpyron replied:
Yes... but only for a block that we already inlined once - and removing one jump to it will only make it *more* likely to inline it again afterwards, so probably fine...
Then again I could probably remove this and the loop below for now without any effect as well...

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,216 @@
+			AssemblyItem const& nextItem = *next(it);
+			if (item.type() == PushTag && nextItem == Instruction::JUMP)
+				if (auto* inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
+					if (auto exitJump = shouldInline(item.data(), nextItem, *inlinableBlock))
+					{
+						newItems += ranges::views::drop_last(inlinableBlock->items, 1);
+						newItems.emplace_back(*exitJump);
```
Wouldn't it be better to just return the jump type?
```suggestion
					if (auto exitJumpType = shouldInline(item.data(), nextItem, *inlinableBlock))
					{
						newItems += inlinableBlock->items;
						newItems.back().setJumpTypE(exitJump);
```

ekpyron replied:
Hm... unfortunately that's again only defined inline in ``AssemblyItem``, so I can't forward-declare it... maybe just properly including ``AssemblyItem`` there isn't too bad in the end...

chriseth commented on test/cmdlineTests/optimizer_BlockDeDuplicator/output:
```diff
@@ -66,24 +66,24 @@ sub_0: assembly {
       stop
+        /* "optimizer_BlockDeDuplicator/input.sol":108:133  function fun_() public {} */
+    tag_4:
+      jump(tag_3)
```
This could actually be replaced by a single `stop`...

ekpyron replied:
Yes - in general we can consider also inlining ``pushtag(tag) jump`` when we see ``tag: ... stop`` or ``tag: ... invalid`` or ``tag: ... revert``... it'd need a different heuristics, though.

ekpyron replied:
Of course the size of the block to be inlined being smaller than the ``pushtag(tag) jump`` itself will always be good...

chriseth replied:
I think it would be useful, even if again not that nice for the debugger. But maybe we can do that in another PR.

ekpyron commented:
I just updated this to split the tag data into subId and tag everywhere and to only ever consider tags referring to the local subassembly (I think we could never have inlined across subassemblys anyways and this way push counting and tag tracking should be more robust).

I also pass in the externally referenced tags, s.t. we can consider them in the heuristics.

Also I addressed most of the other comments I think, so mainly missing are tests and docs I hope.

hrkrshnn commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,256 @@
+
+	// Only consider blocks that end in a JUMP for now. This can e.g. be extended to include transaction terminating
+	// instructions as well in the future.
+	if (_items.back() != Instruction::JUMP)
```
```suggestion
	if (_items.back() != Instruction::JUMP && _items.back().getJumpType() == AssemblyItem::JumpType::OutOfFunction)
```
Can't we check the jump type early, or am I missing something here?

ekpyron replied:
Only because we only inline blocks ending in ``OutOfFunction`` right now, does not mean that we could not inline other blocks as well, so I'd say we collect all candidates here right now already and only choose to inline specific jump type combinations later. We can at any time extend ``Inliner::shouldInline`` to include more combinations. And it's not like this will be performance critical - all of this is extremely fast anyways.

hrkrshnn replied:
Why not remove the check on the jumptype when we are extending the inliner?

hrkrshnn replied:
But I'm fine with both options.

ekpyron replied:
Well, why build things in such a way that you'll need to change two places later instead of one :-)? It's not like the check in ``shouldInline`` would vanish, because the other half of the jump type we only know there anyways. So I'd say the split makes sense as it is. If it's a bother, I can also remove this preselection entirely and check everything in ``shouldInline`` instead...

chriseth commented:
Can you squash "fix windows build"?

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,256 @@
+u256 executionCost(RangeType&& _itemRange, langutil::EVMVersion _evmVersion)
+{
+	GasMeter gasMeter{std::make_shared<KnownState>(), _evmVersion};
+	auto gasConsumption = ranges::accumulate(std::forward<RangeType>(_itemRange) | ranges::views::transform(
```
What is the purpose of the `forward` here?

ekpyron replied:
It's not really required here - it's just idiomatic to write generic functions like this, so I do it without thinking about it or needing to think about it :-).

ekpyron replied:
But I can remove it here.

chriseth commented on libevmasm/Inliner.cpp:
```diff
@@ -0,0 +1,256 @@
+			lastTag.reset();
+		}
+
+		if (item.type() == Tag)
```
We could assert that this is always a local tag.

ekpyron commented:
@haltman-at @yann300 So as you can see we merged this yesterday, so tonight's nightly release should include this in case you want to have another look before it's released.

The final version is restricted to only inline the "jump-in-jump-out" case, i.e. only entire functions will be inlined for the time being (we will extend this sooner or later, though, including to cases like https://github.com/ethereum/solidity/pull/10761#discussion_r570258200).

We decided to go ahead without first defining new debugging output, since we noticed that the other steps in the optimizer will already ignore the fact that jumps have jump labels and may remove jumps into- or out-of- functions already, so we're not introducing a new problem with this PR, but we're "just" making an existing problem worse - so given that and the fact that the gas savings of this can be significant and it is required for zero-cost workarounds for https://github.com/ethereum/solidity/issues/10698, we thought it worthwhile to merge now.
https://github.com/ethereum/solidity/pull/10929 also added some (minimal) documentation about this.

By now it seems clear that we want to define debugging data that specifies function entries and exits on a more fine-grained level, but it's not yet entirely clear what information we want exactly and in what form (Do we *only* want something equivalent to jump labels on opcode level? Do we also want information about the function entry being "complete", i.e. the arguments and return values being initialized on stack? In general what do we do if opcodes shared among several functions is collapsed? How best to try and make this robust against potential future optimizations? Etc.) - we have a forum topic around the topic of debugging information in general (see https://forum.soliditylang.org/t/debugging-information/77) and can hopefully come up with something soon.
